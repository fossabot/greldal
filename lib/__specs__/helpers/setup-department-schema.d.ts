import * as Knex from "knex";
export declare const setupDepartmentSchema: (knex: Knex) => Promise<void>;
export declare const teardownDepartmentSchema: (knex: Knex) => Promise<void>;
