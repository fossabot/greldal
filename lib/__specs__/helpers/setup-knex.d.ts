import Knex from "knex";
export declare const getConnectionString: () => string | undefined;
export declare const setupKnex: () => Knex;
