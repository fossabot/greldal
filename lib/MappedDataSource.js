"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const errors_1 = require("./errors");
const lodash_1 = require("lodash");
const t = __importStar(require("io-ts"));
const debug_1 = __importDefault(require("debug"));
const MappedField_1 = require("./MappedField");
const MappedAssociation_1 = require("./MappedAssociation");
const inflection_1 = require("inflection");
const graphql_type_mapper_1 = require("./graphql-type-mapper");
const connector_1 = require("./connector");
const utils_1 = require("./utils");
const ReverseMapper_1 = require("./ReverseMapper");
const debug = debug_1.default("greldal:MappedDataSource");
class MappedDataSource {
    constructor(mapping) {
        this.mapping = mapping;
        this.fields = lodash_1.transform(mapping.fields, (result, fieldMapping, name) => {
            result[name] = new MappedField_1.MappedField(this, name, fieldMapping);
        }, {});
        this.associations = lodash_1.transform(mapping.associations, (result, associationMapping, name) => {
            result[name] = new MappedAssociation_1.MappedAssociation(this, name, associationMapping);
        }, {});
        debug("Mapped data source with name: %s, fields: %O, associations: %O", this.mappedName, Object.keys(this.fields), Object.keys(this.associations));
        if (mapping.connector) {
            connector_1.assertSupportedConnector(mapping.connector);
        }
    }
    get connector() {
        return connector_1.assertConnectorConfigured(this.mapping.connector || connector_1.globalConnector);
    }
    rootQuery(aliasHierarchyVisitor) {
        if (this.mapping.rootQuery)
            return this.mapping.rootQuery(aliasHierarchyVisitor);
        return aliasHierarchyVisitor
            ? this.connector(`${this.storedName} as ${aliasHierarchyVisitor.alias}`)
            : this.connector(this.storedName);
    }
    get mappedName() {
        return lodash_1.isString(this.mapping.name)
            ? lodash_1.upperFirst(lodash_1.camelCase(inflection_1.singularize(this.mapping.name)))
            : this.mapping.name.mapped;
    }
    get shallowMappedName() {
        return `Shallow${this.mappedName}`;
    }
    get storedName() {
        return lodash_1.isString(this.mapping.name)
            ? lodash_1.snakeCase(inflection_1.pluralize(this.mapping.name))
            : this.mapping.name.stored;
    }
    get storedColumnNames() {
        return lodash_1.transform(this.fields, (result, field, name) => {
            if (field.isMappedFromColumn) {
                result.push(field.sourceColumn);
            }
        }, []);
    }
    get shallowRecordProps() {
        return lodash_1.transform(this.mapping.fields, (result, f, name) => {
            result[name] = f.type;
        }, {});
    }
    get associationProps() {
        const result = {};
        lodash_1.forEach(this.associations, (association, name) => {
            result[name] = association.target.recordType;
        });
        return lodash_1.transform(this.associations, (result, association, name) => {
            result[name] = association.target.recordType;
        }, {});
    }
    get recordProps() {
        return Object.assign({}, this.shallowRecordProps, this.associationProps);
    }
    get shallowRecordType() {
        return t.type(this.shallowRecordProps);
    }
    get recordType() {
        return t.type(this.recordProps);
    }
    get ShallowRecordType() {
        throw errors_1.getTypeAccessorError("ShallowRecordType", "MappedDataSource");
    }
    get RecordType() {
        throw errors_1.getTypeAccessorError("NestedRecordType", "MappedDataSource");
    }
    get MappingType() {
        throw errors_1.getTypeAccessorError("MappingType", "MappedDataSource");
    }
    get defaultOutputType() {
        return graphql_type_mapper_1.deriveDefaultOutputType(this);
    }
    get defaultShallowInputType() {
        debug("Deriving default shallow input type for: ", this.mappedName);
        return graphql_type_mapper_1.deriveDefaultShallowInputType(this);
    }
    get defaultShallowOutputType() {
        return graphql_type_mapper_1.deriveDefaultShallowOutputType(this);
    }
    mapEntities(entities) {
        return entities.map(entity => lodash_1.reduce(entity, (result, val, key) => {
            const field = this.fields[key];
            if (field) {
                result[field.sourceColumn] = val;
            }
            return result;
        }, {}));
    }
    mapResults(storeParams, rows) {
        return new ReverseMapper_1.ReverseMapper(this, storeParams).reverseMap(rows);
    }
    shallowMapResults(rows) {
        rows.map(row => {
            const mappedRow = {};
            for (const [key, field] of Object.entries(this.fields)) {
                mappedRow[field.mappedName] = field.getValue(row);
            }
            return mappedRow;
        });
    }
}
__decorate([
    utils_1.MemoizeGetter,
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [])
], MappedDataSource.prototype, "mappedName", null);
__decorate([
    utils_1.MemoizeGetter,
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [])
], MappedDataSource.prototype, "shallowMappedName", null);
__decorate([
    utils_1.MemoizeGetter,
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [])
], MappedDataSource.prototype, "storedName", null);
__decorate([
    utils_1.MemoizeGetter,
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [])
], MappedDataSource.prototype, "storedColumnNames", null);
__decorate([
    utils_1.MemoizeGetter,
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [])
], MappedDataSource.prototype, "shallowRecordProps", null);
__decorate([
    utils_1.MemoizeGetter,
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [])
], MappedDataSource.prototype, "associationProps", null);
__decorate([
    utils_1.MemoizeGetter,
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [])
], MappedDataSource.prototype, "recordProps", null);
__decorate([
    utils_1.MemoizeGetter,
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [])
], MappedDataSource.prototype, "shallowRecordType", null);
__decorate([
    utils_1.MemoizeGetter,
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [])
], MappedDataSource.prototype, "recordType", null);
__decorate([
    utils_1.MemoizeGetter,
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [])
], MappedDataSource.prototype, "defaultOutputType", null);
__decorate([
    utils_1.MemoizeGetter,
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [])
], MappedDataSource.prototype, "defaultShallowInputType", null);
__decorate([
    utils_1.MemoizeGetter,
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [])
], MappedDataSource.prototype, "defaultShallowOutputType", null);
exports.MappedDataSource = MappedDataSource;
exports.mapDataSource = (mapping) => new MappedDataSource(mapping);
//# sourceMappingURL=MappedDataSource.js.map