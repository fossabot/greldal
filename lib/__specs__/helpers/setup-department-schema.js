"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.setupDepartmentSchema = (knex) => __awaiter(this, void 0, void 0, function* () {
    yield knex.schema.createTable("tags", t => {
        t.increments("id");
        t.string("name");
    });
    yield knex.schema.createTable("departments", t => {
        t.increments("id");
        t.string("name");
    });
    yield knex.schema.createTable("products", t => {
        t.increments("id");
        t.integer("department_id")
            .references("id")
            .inTable("departments")
            .notNullable();
        t.string("name");
    });
    yield knex.schema.createTable("product_tag_associators", t => {
        t.increments("id");
        t.integer("product_id")
            .references("id")
            .inTable("products")
            .notNullable();
        t.integer("tag_id")
            .references("id")
            .inTable("tags")
            .notNullable();
    });
});
exports.teardownDepartmentSchema = (knex) => __awaiter(this, void 0, void 0, function* () {
    yield knex.schema.dropTable("departments");
    yield knex.schema.dropTable("products");
    yield knex.schema.dropTable("tags");
    yield knex.schema.dropTable("product_tag_associators");
});
//# sourceMappingURL=setup-department-schema.js.map