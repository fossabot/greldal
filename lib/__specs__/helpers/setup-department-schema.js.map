{"version":3,"file":"setup-department-schema.js","sourceRoot":"","sources":["../../../src/__specs__/helpers/setup-department-schema.ts"],"names":[],"mappings":";;;;;;;;;;AAEa,QAAA,qBAAqB,GAAG,CAAO,IAAU,EAAE,EAAE;IACtD,MAAM,IAAI,CAAC,MAAM,CAAC,WAAW,CAAC,MAAM,EAAE,CAAC,CAAC,EAAE;QACtC,CAAC,CAAC,UAAU,CAAC,IAAI,CAAC,CAAC;QACnB,CAAC,CAAC,MAAM,CAAC,MAAM,CAAC,CAAC;IACrB,CAAC,CAAC,CAAC;IACH,MAAM,IAAI,CAAC,MAAM,CAAC,WAAW,CAAC,aAAa,EAAE,CAAC,CAAC,EAAE;QAC7C,CAAC,CAAC,UAAU,CAAC,IAAI,CAAC,CAAC;QACnB,CAAC,CAAC,MAAM,CAAC,MAAM,CAAC,CAAC;IACrB,CAAC,CAAC,CAAC;IACH,MAAM,IAAI,CAAC,MAAM,CAAC,WAAW,CAAC,UAAU,EAAE,CAAC,CAAC,EAAE;QAC1C,CAAC,CAAC,UAAU,CAAC,IAAI,CAAC,CAAC;QACnB,CAAC,CAAC,OAAO,CAAC,eAAe,CAAC;aACrB,UAAU,CAAC,IAAI,CAAC;aAChB,OAAO,CAAC,aAAa,CAAC;aACtB,WAAW,EAAE,CAAC;QACnB,CAAC,CAAC,MAAM,CAAC,MAAM,CAAC,CAAC;IACrB,CAAC,CAAC,CAAC;IACH,MAAM,IAAI,CAAC,MAAM,CAAC,WAAW,CAAC,yBAAyB,EAAE,CAAC,CAAC,EAAE;QACzD,CAAC,CAAC,UAAU,CAAC,IAAI,CAAC,CAAC;QACnB,CAAC,CAAC,OAAO,CAAC,YAAY,CAAC;aAClB,UAAU,CAAC,IAAI,CAAC;aAChB,OAAO,CAAC,UAAU,CAAC;aACnB,WAAW,EAAE,CAAC;QACnB,CAAC,CAAC,OAAO,CAAC,QAAQ,CAAC;aACd,UAAU,CAAC,IAAI,CAAC;aAChB,OAAO,CAAC,MAAM,CAAC;aACf,WAAW,EAAE,CAAC;IACvB,CAAC,CAAC,CAAC;AACP,CAAC,CAAA,CAAC;AAEW,QAAA,wBAAwB,GAAG,CAAO,IAAU,EAAE,EAAE;IACzD,MAAM,IAAI,CAAC,MAAM,CAAC,SAAS,CAAC,aAAa,CAAC,CAAC;IAC3C,MAAM,IAAI,CAAC,MAAM,CAAC,SAAS,CAAC,UAAU,CAAC,CAAC;IACxC,MAAM,IAAI,CAAC,MAAM,CAAC,SAAS,CAAC,MAAM,CAAC,CAAC;IACpC,MAAM,IAAI,CAAC,MAAM,CAAC,SAAS,CAAC,yBAAyB,CAAC,CAAC;AAC3D,CAAC,CAAA,CAAC"}