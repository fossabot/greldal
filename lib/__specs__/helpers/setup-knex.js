"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const knex_1 = __importDefault(require("knex"));
const path_1 = __importDefault(require("path"));
const tmp_1 = __importDefault(require("tmp"));
const assert_1 = __importDefault(require("assert"));
const debug_1 = __importDefault(require("debug"));
const debug = debug_1.default("greldal:setup-knex");
exports.getConnectionString = () => {
    const connectionString = process.env.DB_CONNECTION_STRING;
    assert_1.default(connectionString, "Expected DB_CONNECTION_STRING environment variable to be defined");
    return connectionString;
};
exports.setupKnex = () => {
    const isCI = !!process.env.CI;
    let config;
    switch (process.env.DB) {
        case "mysql":
            if (isCI) {
                config = {
                    client: "mysql",
                    connection: {
                        socketPath: path_1.default.join(process.env.HOME, ".my.cnf"),
                        user: "travis",
                        database: "greldal_test",
                    },
                };
            }
            else {
                config = {
                    client: "mysql",
                    connection: exports.getConnectionString(),
                };
            }
        case "pg":
            if (isCI) {
                config = {
                    client: "pg",
                    connection: {
                        user: "postgres",
                        database: "greldal_test",
                    },
                };
            }
            else {
                config = {
                    client: "pg",
                    connection: exports.getConnectionString(),
                };
            }
        case "sqlite3":
        default:
            const sqliteFile = tmp_1.default.fileSync();
            config = {
                client: "sqlite3",
                connection: {
                    filename: sqliteFile.name,
                },
                useNullAsDefault: true,
            };
    }
    debug("Connecting to knex using configuration: %O", config);
    assert_1.default(config, "Failed to configure database for the test suite. You may need to update above configuration");
    return knex_1.default(config);
};
//# sourceMappingURL=setup-knex.js.map