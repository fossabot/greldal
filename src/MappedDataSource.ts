import { getTypeAccessorError } from "./errors";
import { Memoize } from "lodash-decorators";
import { Mapped, TypeGuard, Dict, MaybeMapped, MaybeArray, NNil } from "./util-types";
import { isString, transform, camelCase, upperFirst, snakeCase } from "lodash";
import * as t from "io-ts";
import { GraphQLInputType, GraphQLOutputType } from "graphql";
import { FieldMapping, MappedField } from "./MappedField";
import { AssociationMapping, MappedAssociation } from "./MappedAssociation";
import { singularize, pluralize } from "inflection";
import { deriveDefaultShallowOutputType, deriveDefaultOutputType, deriveDefaultShallowInputType } from "./graphql-type-mapper";

export interface DataSourceMapping {
    name: MaybeMapped<string>;
    description?: string;
    fields?: Dict<FieldMapping<any, any>>
    associations?: Dict<AssociationMapping<any>>
}

type ShallowRecordType<T extends DataSourceMapping> = {
    [K in keyof T["fields"]]: t.TypeOf<NNil<T["fields"]>[K]["type"]>
}

type NestedRecordType<T extends DataSourceMapping> = ShallowRecordType<T> & {
    [K in keyof T["associations"]]: ReturnType<NNil<T["associations"]>[K]["from"]>["NestedRecordType"]
}

export class MappedDataSource<T extends DataSourceMapping = any> {
    fields: { [K in keyof T["fields"]]: MappedField<NNil<T["fields"]>[K]> };
    associations: { [K in keyof T["associations"]]: MappedAssociation<NNil<T["associations"]>[K]> };

    constructor(private mapping: T) {
        this.fields = transform(
            mapping.fields!,
            (result, fieldMapping, name) => {
                result[name] = new MappedField(this, name, fieldMapping);
            }, {}
        ) as any;
        this.associations = transform(
            mapping.associations!,
            (result, associationMapping, name) => {
                result[name] = new MappedAssociation(this, name, associationMapping);
            }, {}
        ) as any;
    }

    @Memoize
    get mappedName() {
        return (isString as TypeGuard<string>)(this.mapping.name)
            ? upperFirst(camelCase(singularize(this.mapping.name)))
            : this.mapping.name.mapped;
    }

    @Memoize
    get storedName() {
        return (isString as TypeGuard<string>)(this.mapping.name)
            ? snakeCase(pluralize(this.mapping.name))
            : this.mapping.name.stored;
    }

    get ShallowRecordType(): ShallowRecordType<T> {
        throw getTypeAccessorError('ShallowRecordType', 'MappedDataSource');
    }

    get NestedRecordType(): NestedRecordType<T> {
        throw getTypeAccessorError('NestedRecordType', 'MappedDataSource');
    }

    @Memoize
    get defaultOutputType(): GraphQLOutputType {
        return deriveDefaultOutputType(this);
    }

    @Memoize
    get defaultShallowInputType(): GraphQLInputType {
        return deriveDefaultShallowInputType(this);
    }

    @Memoize
    get defaultShallowOutputType(): GraphQLOutputType {
        return deriveDefaultShallowOutputType(this)
    }
}

export const dataSource = <T extends DataSourceMapping>(mapping: T) =>
    new MappedDataSource<T>(mapping);
