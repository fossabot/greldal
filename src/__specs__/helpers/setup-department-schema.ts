import * as Knex from "knex";

export const setupDepartmentSchema = async (knex: Knex) => {
    await knex.schema.createTable("tags", t => {
        t.increments("id");
        t.string("name");
    });
    await knex.schema.createTable("departments", t => {
        t.increments("id");
        t.string("name");
    });
    await knex.schema.createTable("products", t => {
        t.increments("id");
        t.integer("department_id")
            .references("id")
            .inTable("departments")
            .notNullable();
        t.string("name");
    });
    await knex.schema.createTable("product_tag_associators", t => {
        t.increments("id");
        t.integer("product_id")
            .references("id")
            .inTable("products")
            .notNullable();
        t.integer("tag_id")
            .references("id")
            .inTable("tags")
            .notNullable();
    });
};

export const teardownDepartmentSchema = async (knex: Knex) => {
    await knex.schema.dropTable("departments");
    await knex.schema.dropTable("products");
    await knex.schema.dropTable("tags");
    await knex.schema.dropTable("product_tag_associators");
};
