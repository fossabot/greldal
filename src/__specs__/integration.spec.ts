import Knex from "knex";
import { useDatabaseConnector, mapDataSource, mapSchema, types, operationPresets } from "..";
import { setupKnex } from "./helpers/setup-knex";
import { GraphQLID, printSchema, graphql, GraphQLSchema } from "graphql";
import { MappedDataSource } from "../MappedDataSource";

let knex: Knex;

beforeAll(() => {
    knex = setupKnex();
    useDatabaseConnector(knex);
});

describe("Conventionally mapped data source", () => {
    let mappedDataSource: MappedDataSource, generatedSchema: GraphQLSchema;
    beforeAll(async () => {
        await knex.schema.createTable("users", t => {
            t.increments("id");
            t.string("name");
        });
        mappedDataSource = mapDataSource({
            name: "User",
            fields: {
                id: {
                    type: types.number,
                    to: GraphQLID,
                },
                name: {
                    type: types.string,
                },
            },
        });
        generatedSchema = mapSchema(operationPresets.all(mappedDataSource));
        await knex("users").insert([{ id: 1, name: "Lorefnon" }, { id: 2, name: "Gandalf" }]);
    });
    afterAll(async () => {
        await knex.schema.dropTable("users");
    });
    test("generated schema", () => {
        expect(printSchema(generatedSchema)).toMatchSnapshot();
    });
    test("singular query operation without params", async () => {
        const r1 = await graphql(
            generatedSchema,
            `
                query {
                    findOneUser(where: {}) {
                        id
                        name
                    }
                }
            `,
        );
        expect(r1.errors).not.toBeDefined();
        expect(r1).toMatchSnapshot();
    });
    test("singular query operation with params", async () => {
        const r2 = await graphql(
            generatedSchema,
            `
                query {
                    findOneUser(where: { id: 2 }) {
                        id
                        name
                    }
                }
            `,
        );
        expect(r2.errors).not.toBeDefined();
        expect(r2).toMatchSnapshot();
    });
    test("singular query operation for non-existent row", async () => {
        const r3 = await graphql(
            generatedSchema,
            `
                query {
                    findOneUser(where: { id: 10 }) {
                        id
                        name
                    }
                }
            `,
        );
        expect(r3.errors).not.toBeDefined();
        expect(r3.data!.findOneUser).toBeNull();
    });
    test("batch query operation without args", async () => {
        const r4 = await graphql(
            generatedSchema,
            `
                query {
                    findManyUsers(where: {}) {
                        id
                        name
                    }
                }
            `,
        );
        expect(r4.errors).not.toBeDefined();
        expect(r4).toMatchSnapshot();
    });
    test("batch query operations with arguments", async () => {
        const r5 = await graphql(
            generatedSchema,
            `
                query {
                    findManyUsers(where: { id: 1 }) {
                        id
                        name
                    }
                }
            `,
        );
        expect(r5.errors).not.toBeDefined();
        expect(r5).toMatchSnapshot();
    });
});

describe("Data source mapped as per custom configuration", () => {
    let mappedDataSource: MappedDataSource, generatedSchema: GraphQLSchema;
    beforeAll(async () => {
        await knex.schema.createTable("customers", t => {
            t.increments("pk");
            t.string("first_name");
            t.string("last_name");
        });
        mappedDataSource = mapDataSource({
            name: {
                mapped: "User",
                stored: "customers",
            },
            fields: {
                id: {
                    sourceColumn: "pk",
                    type: types.string,
                    to: {
                        input: GraphQLID,
                        output: GraphQLID,
                    },
                },
                firstName: {
                    sourceColumn: "first_name",
                    type: types.string,
                },
                lastName: {
                    sourceColumn: "last_name",
                    type: types.string,
                },
            },
        });
        generatedSchema = mapSchema(operationPresets.all(mappedDataSource));
        await knex("customers").insert([
            { pk: 1, first_name: "John", last_name: "Doe" },
            { pk: 2, first_name: "Jane", last_name: "Doe" },
        ]);
    });
    afterAll(async () => {
        await knex.schema.dropTable("customers");
    });
    test("generated schema", () => {
        expect(printSchema(generatedSchema)).toMatchSnapshot();
    });
    test("singular query operation", async () => {
        const r1 = await graphql(generatedSchema, "query { findOneUser(where: {}) { id, firstName, lastName }}");
        expect(r1.errors).not.toBeDefined();
        expect(r1).toMatchSnapshot();
        const r2 = await graphql(generatedSchema, "query { findOneUser(where: {id: 2}) { id, firstName, lastName }}");
        expect(r2.errors).not.toBeDefined();
        expect(r2).toMatchSnapshot();
    });
    test("batch query operation", async () => {
        const r3 = await graphql(generatedSchema, "query { findManyUsers(where: {}) { id, firstName, lastName }}");
        expect(r3.errors).not.toBeDefined();
        expect(r3).toMatchSnapshot();
    });
});

describe.only("Data sources associated by joins", () => {
    let tags: MappedDataSource, products: MappedDataSource, departments: MappedDataSource;
    let generatedSchema: GraphQLSchema;
    beforeAll(async () => {
        await knex.schema.createTable("tags", t => {
            t.increments("id");
            t.string("name");
        });
        await knex.schema.createTable("departments", t => {
            t.increments("id");
            t.string("name");
        });
        await knex.schema.createTable("products", t => {
            t.increments("id");
            t.integer("department_id")
                .references("id")
                .inTable("departments")
                .notNullable();
            t.string("name");
        });
        await knex.schema.createTable("product_tag_associators", t => {
            t.increments("id");
            t.integer("product_id")
                .references("id")
                .inTable("products")
                .notNullable();
            t.integer("tag_id")
                .references("id")
                .inTable("tags")
                .notNullable();
        });
        await knex("tags").insert([{ name: "imported" }, { name: "third-party" }]);
        await knex("departments").insert([{ name: "textile" }, { name: "heavy goods" }]);
        await knex("products").insert([
            { name: "silk gown", department_id: 1 },
            { name: "steel welding machine", department_id: 2 },
        ]);
        await knex("product_tag_associators").insert([
            { product_id: 1, tag_id: 1 },
            { product_id: 2, tag_id: 2 },
            { product_id: 2, tag_id: 1 },
        ]);
        const fields = {
            id: {
                type: types.number,
                to: GraphQLID,
            },
            name: {
                type: types.string,
            },
        };
        tags = mapDataSource({
            name: "Tag",
            fields,
            associations: {
                products: {
                    target: () => products,
                    join: (queryBuilder, ahv) => {
                        const ptaVisitor = ahv.visit('product_tag_associators');
                        const productsVisitor = ptaVisitor.visit('products');
                        queryBuilder
                            .leftOuterJoin(
                                `product_tag_associators as ${ptaVisitor.alias}`,
                                `${ptaVisitor.alias}.tag_id`,
                                `${ahv.alias}.id`,
                            )
                            .leftOuterJoin(
                                `products as ${productsVisitor.alias}`,
                                `${productsVisitor.alias}.id`,
                                `${ptaVisitor.alias}.product_id`,
                            );
                        return productsVisitor;
                    },
                    singular: false,
                },
            },
        });
        products = mapDataSource({
            name: "Product",
            fields,
            associations: {
                department: {
                    target: () => departments,
                    singular: true,
                    join: "leftOuterJoin",
                    associatorColumns: {
                        inSource: "department_id",
                        inRelated: "id",
                    },
                },
            },
        });
        departments = mapDataSource({
            name: "Department",
            fields,
            associations: {
                products: {
                    target: () => products,
                    singular: false,
                    join: "leftOuterJoin",
                    associatorColumns: {
                        inSource: "id",
                        inRelated: "department_id",
                    },
                },
            },
        });
        generatedSchema = mapSchema([
            ...operationPresets.all(tags),
            ...operationPresets.all(departments),
            ...operationPresets.all(products),
        ]);
    });
    test("generated schema", () => {
        expect(printSchema(generatedSchema)).toMatchSnapshot();
    });
    test("query operations involving auto-inferred binary joins", async () => {
        const r1 = await graphql(
            generatedSchema,
            `
                query {
                    findOneProduct(where: {}) {
                        id
                        name
                        department(where: {}) {
                            id
                            name
                        }
                    }
                }
            `,
        );
        expect(r1).toMatchSnapshot();
        const r2 = await graphql(
            generatedSchema,
            `
                query {
                    findOneDepartment(where: { id: 1 }) {
                        id
                        name
                        products(where: {}) {
                            id
                            name
                        }
                    }
                }
            `,
        );
        expect(r2).toMatchSnapshot();
    });
    test("query operations involving user specified complex joins", async () => {
        const r1 = await graphql(
            generatedSchema,
            `
                query {
                    findManyTags(where: {}) {
                        id
                        name
                        products(where: {}) {
                            id
                            name
                        }
                    }
                }
            `,
        );
        expect(r1).toMatchSnapshot();
    });
    afterAll(async () => {
        await knex.schema.dropTable("departments");
        await knex.schema.dropTable("products");
        await knex.schema.dropTable("tags");
    });
});

describe("Data sources linked by pre-fetchable associations", async () => {

});

describe("Data sources linked by post-fetchable associations", async () => {

});
