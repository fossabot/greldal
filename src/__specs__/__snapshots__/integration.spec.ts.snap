// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Conventionally mapped data source batch query operation without args 1`] = `
Object {
  "data": Object {
    "findManyUsers": Array [
      Object {
        "id": "1",
        "name": "Lorefnon",
      },
      Object {
        "id": "2",
        "name": "Gandalf",
      },
    ],
  },
}
`;

exports[`Conventionally mapped data source batch query operations with arguments 1`] = `
Object {
  "data": Object {
    "findManyUsers": Array [
      Object {
        "id": "1",
        "name": "Lorefnon",
      },
    ],
  },
}
`;

exports[`Conventionally mapped data source generated schema 1`] = `
"schema {
  query: query
  mutation: mutation
}

type mutation {
  insertOneUser(entity: UserInput): ShallowUser
  insertManyUsers(entities: [UserInput]): [ShallowUser]
  updateOneUser(where: UserInput!, update: UserInput!): ShallowUser
  updateManyUsers(where: UserInput!, update: UserInput!): [ShallowUser]
  deleteOneUser(where: UserInput!): ShallowUser
  deleteManyUsers(where: UserInput!): [ShallowUser]
}

type query {
  findOneUser(where: UserInput!): User
  findManyUsers(where: UserInput!): [User]
}

type ShallowUser {
  id: ID
  name: String
}

type User {
  id: ID
  name: String
}

input UserInput {
  id: ID
  name: String
}
"
`;

exports[`Conventionally mapped data source singular query operation with params 1`] = `
Object {
  "data": Object {
    "findOneUser": Object {
      "id": "2",
      "name": "Gandalf",
    },
  },
}
`;

exports[`Conventionally mapped data source singular query operation without params 1`] = `
Object {
  "data": Object {
    "findOneUser": Object {
      "id": "1",
      "name": "Lorefnon",
    },
  },
}
`;

exports[`Data source mapped as per custom configuration generated schema 1`] = `
"schema {
  query: query
  mutation: mutation
}

type mutation {
  insertOneUser(entity: UserInput): ShallowUser
  insertManyUsers(entities: [UserInput]): [ShallowUser]
  updateOneUser(where: UserInput!, update: UserInput!): ShallowUser
  updateManyUsers(where: UserInput!, update: UserInput!): [ShallowUser]
  deleteOneUser(where: UserInput!): ShallowUser
  deleteManyUsers(where: UserInput!): [ShallowUser]
}

type query {
  findOneUser(where: UserInput!): User
  findManyUsers(where: UserInput!): [User]
}

type ShallowUser {
  id: ID
  firstName: String
  lastName: String
}

type User {
  id: ID
  firstName: String
  lastName: String
}

input UserInput {
  id: ID
  firstName: String
  lastName: String
}
"
`;

exports[`Data source mapped as per custom configuration singular query operation 1`] = `
Object {
  "data": Object {
    "findOneUser": Object {
      "firstName": "John",
      "id": "1",
      "lastName": "Doe",
    },
  },
}
`;

exports[`Data source mapped as per custom configuration singular query operation 2`] = `
Object {
  "data": Object {
    "findOneUser": Object {
      "firstName": "Jane",
      "id": "2",
      "lastName": "Doe",
    },
  },
}
`;
